README.md
# What's jBATIS
The jBATIS persistence framework is a succeeding framework for iBATIS Sql Map framework which was stopped at version 2.3.4.726.

# Features & Improvements

 * cglib and asm refined and embedded.
 * jdbc null value unboxed as 0.
 * improved batch: auto tx, auto batch size and call procedure in batch
 * SIMPLE datasource fixed and refined for industrial-scale applications.
 * classpath:/ibatis.ini for global properties
 * 2.4 dtd and dynamic sql tag alias added
 * new APIs in org.ibatis.client.SqlMapClient and its super interfaces.
 * new Spring integration support.
 * jBATIS Persistence APIs
 * cache support refined and 3rd lib refreshed.
 * dialect and paging query support for modern databases.
 
# Maven artifact
```xml
  <dependency>
    <groupId>com.dukeware</groupId>
    <artifactId>jbatis</artifactId>
    <version>2.4.8</version>
  </dependency>
```

# User Guide

 see doc/jbatis-guide.md

doc/changelog.txt
2018-10-30  2.4.7 again! bugfix: npe if no ibatis.ini

2018-10-10  bugfix: count paging query for oracle 12c+

2018-6-4    bugfix: NPE while batch sql stats
            refine code and performance

2018-5-21   refine sql check

2018-5-15   refine performance and SimpleDataSource lock

2018-5-9    refine dtd 2.4: add canBatch and noBatch attr
            fixbug: method override lost in ProcedureStatement and CachingStatement
            refine SimpleDataSource lock and performance
            add Pool.CommitOnReturn parameter, default false
            sync cglib caches
            refine ArrayMap and erase unused code

2018-4-24   refine performance

2018-4-24   refine batch sql stats

2018-4-23   fixbug: now # can in ''
            refine performance

2018-4-20   jbatis-guide.md released.
            refine code consistency with the guide
            refine dynamic tag processing performance
            refine sql stats code
            fixbug: list data exchange fail when in quick <iterate>

2018-4-13   check_sql to check <select> remapResults flag
            refine dynamic sql generating
            refine duplicated statement check
            refine duplicated sql map resource log
            sync refine from https://github.com/mybatis/ibatis-2
            some code fixes.

2018-3-29   spring 4.x and 5.x verified.

2018-2-12   fixbug: paging query for sqlserver 2012+
            paging query for oracle 12c+ and db2 9.7+
            refine SimpleDataSource

2018-1-10   refine sql text process

2018-1-5    add page dialect for sql server 2012+
            fixbug: now $ can in ''
            refine perfermance

2018-1-4    update to jbatis 2.4.7
            refine page query
            bugfix: page query via entity manager
            refine perfermance
            add mysql page dialect
            add sql lexer and refine LogR
            add postgresql page dialect

2017-12-20  refine duplicated statement judgement

2017-12-6   update to jbatis 2.4.6
            refine <iterate> perfermance and bugfix
            flush cache at batch end for perfermance
            refine some log
            refine LogR

2017-12-1   <iterate> add Set support and refine perfermance
            refine fast field and fast method
            cglib_cache_size default 2048 in ibatis.ini
            split atomikos to 4.0.5

2017-11-16  refine perfermance when no parameter
            erase dom probe

2017-11-7   parameter handler: _like_ support
            refine 2.4 dtd

2017-9-24   refine cache notification

2017-4-19,  rename to jbatis 2.4.5
            xa support

2017-4-1,   also use stream to parse <sqlMap> if Resources.charset is null

2016-12-29, None cache for special usage
            refine DaoSupport

2016-12-12, support <settings defaultCacheModelType="*" forceCacheModelType="*"/>
            no sync flush if no memory.mcast.enable=true in ibatis.ini
            refine MEMORY cache
            support ${key:defaultValue} in xml

2016-10-27, refine SimpleDataSource lock
            cache_flush_monitor in stats file: <cache-id>.onFlush=error|info

2016-10-14, cglib enhanced and bugfix from 88 or ealier

---------------------------------------------------------------
1, Impl the jdbc4 api.
2, Support setParameter() for non-type null value.
 - (fixed by 2.3.5)
3, Ignore the primitive null value while processing ResultSet.
4, Add cached rowset impl for result exporting.
5, Add dialect support.
6, Update to 2.3.5
7, Improve the batch processing.
8, Add auto tx in batch.
9, cglib added. 
10, unbox null value as zero
11, change default idle conn nummber from 5 to 2
12, set global props into ds
13, support to ping idle conn for SIMPLE data source
14, support to erase idle conn for SIMPLE data source
15, refine the error context output
16, read default global properties from classpath:/ibatis.ini
17, bulk with fileds
18, refine the log of SimpleDataSource
19, add sql args to error ctx
20, sleep 15s before close conn pool
21, trim the sql string
22, erase too old conn for SIMPLE data source
23, adjust the error log and pool log
24, refactor cglib and asm, adjust the logs of jdbc pool
25, batch bug fix, start batch error if exists one
26, call procedure in batch
27, support logging slow sql
28, support spring integration
-- Upgrade to 2.4.0 2012-6
29, add dynamic sql tag alias and bugfix-1
30, support debug sql
31, improve query for map
32, keep object id while applying result-map
33, ignore first char's case of bean property name.
34, CacheModel trace log for object retrieved
35, better ehcache support
36, only warn when cache triger missed
37, refine too many result error
38, refine batch error detail
39, better sql debug logging
40, queryForObject() bug fixed from 37
41, better debug info for 30 and 27
42, queryForFirst optimized
43, add $@param$ support
44, getRealDataSource() in org.ibatis.client.SqlMapClient
45, support defaultTransactionIsolation for SIMPLE data source
45, refine status log and tx isolation def
46, refine isValid() for SIMPLE data source
47, fixbug: sqlMap resource lost when using spring mappingLocations
48, refine debug output for date parameter
49, refine batch and debug info
50, fixbug: args debug info
51, fixbug: resultMap <result property="aaa.bbb" column="ccc" />
52, refine exception path
53, refine formatting sql text
54, fixbug: dynamic sql in same batch cause error since #25
55, refine formatting sql text for performance
56, auto warp ' ' before and after prependAttr.trim()
57, support batchSize via startBatch(int)
58, refine exceptions in SIMPLE data source
59, refine: skipResults < 0
60, spring factory bean bug fix.
61, ver 2.4.1: entity api 0.5
62, fix all generic warning
63, add criteria api
64, add settings:jdbcTypeForNull suppport
65, Pool.MaximumCheckoutTime -1 means unlimited.

src/main/java/org/ibatis/asm/Opcodes.java
/***
 * ASM: a very small and fast Java bytecode manipulation framework
 * Copyright (c) 2000-2011 INRIA, France Telecom
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holders nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.ibatis.asm;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.math.BigDecimal;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Defines the JVM opcodes, access flags and array type codes. This interface
 * does not define all the JVM opcodes because some opcodes are automatically
 * handled. For example, the xLOAD and xSTORE opcodes are automatically replaced
 * by xLOAD_n and xSTORE_n opcodes when possible. The xLOAD_n and xSTORE_n
 * opcodes are therefore not defined in this interface. Likewise for LDC,
 * automatically replaced by LDC_W or LDC2_W when necessary, WIDE, GOTO_W and
 * JSR_W.
 * 
 * @author Eric Bruneton
 * @author Eugene Kuleshov
 */
public interface Opcodes {

    // ASM API versions

    int ASM4 = 4 << 16 | 0 << 8 | 0;
    int ASM5 = 5 << 16 | 0 << 8 | 0;

    // versions

    int V1_1 = 3 << 16 | 45;
    int V1_2 = 0 << 16 | 46;
    int V1_3 = 0 << 16 | 47;
    int V1_4 = 0 << 16 | 48;
    int V1_5 = 0 << 16 | 49;
    int V1_6 = 0 << 16 | 50;
    int V1_7 = 0 << 16 | 51;
    int V1_8 = 0 << 16 | 52;

    // access flags

    int ACC_PUBLIC = 0x0001; // class, field, method
    int ACC_PRIVATE = 0x0002; // class, field, method
    int ACC_PROTECTED = 0x0004; // class, field, method
    int ACC_STATIC = 0x0008; // field, method
    int ACC_FINAL = 0x0010; // class, field, method, parameter
    int ACC_SUPER = 0x0020; // class
    int ACC_SYNCHRONIZED = 0x0020; // method
    int ACC_VOLATILE = 0x0040; // field
    int ACC_BRIDGE = 0x0040; // method
    int ACC_VARARGS = 0x0080; // method
    int ACC_TRANSIENT = 0x0080; // field
    int ACC_NATIVE = 0x0100; // method
    int ACC_INTERFACE = 0x0200; // class
    int ACC_ABSTRACT = 0x0400; // class, method
    int ACC_STRICT = 0x0800; // method
    int ACC_SYNTHETIC = 0x1000; // class, field, method, parameter
    int ACC_ANNOTATION = 0x2000; // class
    int ACC_ENUM = 0x4000; // class(?) field inner
    int ACC_MANDATED = 0x8000; // parameter

    // ASM specific pseudo access flags

    int ACC_DEPRECATED = 0x20000; // class, field, method

    // types for NEWARRAY

    int T_BOOLEAN = 4;
    int T_CHAR = 5;
    int T_FLOAT = 6;
    int T_DOUBLE = 7;
    int T_BYTE = 8;
    int T_SHORT = 9;
    int T_INT = 10;
    int T_LONG = 11;

    // tags for Handle

    int H_GETFIELD = 1;
    int H_GETSTATIC = 2;
    int H_PUTFIELD = 3;
    int H_PUTSTATIC = 4;
    int H_INVOKEVIRTUAL = 5;
    int H_INVOKESTATIC = 6;
    int H_INVOKESPECIAL = 7;
    int H_NEWINVOKESPECIAL = 8;
    int H_INVOKEINTERFACE = 9;

    // stack map frame types

    /**
     * Represents an expanded frame. See {@link ClassReader#EXPAND_FRAMES}.
     */
    int F_NEW = -1;

    /**
     * Represents a compressed frame with complete frame data.
     */
    int F_FULL = 0;

    /**
     * Represents a compressed frame where locals are the same as the locals in
     * the previous frame, except that additional 1-3 locals are defined, and
     * with an empty stack.
     */
    int F_APPEND = 1;

    /**
     * Represents a compressed frame where locals are the same as the locals in
     * the previous frame, except that the last 1-3 locals are absent and with
     * an empty stack.
     */
    int F_CHOP = 2;

    /**
     * Represents a compressed frame with exactly the same locals as the
     * previous frame and with an empty stack.
     */
    int F_SAME = 3;

    /**
     * Represents a compressed frame with exactly the same locals as the
     * previous frame and with a single value on the stack.
     */
    int F_SAME1 = 4;

    Integer TOP = new Integer(0);
    Integer INTEGER = new Integer(1);
    Integer FLOAT = new Integer(2);
    Integer DOUBLE = new Integer(3);
    Integer LONG = new Integer(4);
    Integer NULL = new Integer(5);
    Integer UNINITIALIZED_THIS = new Integer(6);

    // opcodes // visit method (- = idem)

    int NOP = 0; // visitInsn
    int ACONST_NULL = 1; // -
    int ICONST_M1 = 2; // -
    int ICONST_0 = 3; // -
    int ICONST_1 = 4; // -
    int ICONST_2 = 5; // -
    int ICONST_3 = 6; // -
    int ICONST_4 = 7; // -
    int ICONST_5 = 8; // -
    int LCONST_0 = 9; // -
    int LCONST_1 = 10; // -
    int FCONST_0 = 11; // -
    int FCONST_1 = 12; // -
    int FCONST_2 = 13; // -
    int DCONST_0 = 14; // -
    int DCONST_1 = 15; // -
    int BIPUSH = 16; // visitIntInsn
    int SIPUSH = 17; // -
    int LDC = 18; // visitLdcInsn
    // int LDC_W = 19; // -
    // int LDC2_W = 20; // -
    int ILOAD = 21; // visitVarInsn
    int LLOAD = 22; // -
    int FLOAD = 23; // -
    int DLOAD = 24; // -
    int ALOAD = 25; // -
    // int ILOAD_0 = 26; // -
    // int ILOAD_1 = 27; // -
    // int ILOAD_2 = 28; // -
    // int ILOAD_3 = 29; // -
    // int LLOAD_0 = 30; // -
    // int LLOAD_1 = 31; // -
    // int LLOAD_2 = 32; // -
    // int LLOAD_3 = 33; // -
    // int FLOAD_0 = 34; // -
    // int FLOAD_1 = 35; // -
    // int FLOAD_2 = 36; // -
    // int FLOAD_3 = 37; // -
    // int DLOAD_0 = 38; // -
    // int DLOAD_1 = 39; // -
    // int DLOAD_2 = 40; // -
    // int DLOAD_3 = 41; // -
    // int ALOAD_0 = 42; // -
    // int ALOAD_1 = 43; // -
    // int ALOAD_2 = 44; // -
    // int ALOAD_3 = 45; // -
    int IALOAD = 46; // visitInsn
    int LALOAD = 47; // -
    int FALOAD = 48; // -
    int DALOAD = 49; // -
    int AALOAD = 50; // -
    int BALOAD = 51; // -
    int CALOAD = 52; // -
    int SALOAD = 53; // -
    int ISTORE = 54; // visitVarInsn
    int LSTORE = 55; // -
    int FSTORE = 56; // -
    int DSTORE = 57; // -
    int ASTORE = 58; // -
    // int ISTORE_0 = 59; // -
    // int ISTORE_1 = 60; // -
    // int ISTORE_2 = 61; // -
    // int ISTORE_3 = 62; // -
    // int LSTORE_0 = 63; // -
    // int LSTORE_1 = 64; // -
    // int LSTORE_2 = 65; // -
    // int LSTORE_3 = 66; // -
    // int FSTORE_0 = 67; // -
    // int FSTORE_1 = 68; // -
    // int FSTORE_2 = 69; // -
    // int FSTORE_3 = 70; // -
    // int DSTORE_0 = 71; // -
    // int DSTORE_1 = 72; // -
    // int DSTORE_2 = 73; // -
    // int DSTORE_3 = 74; // -
    // int ASTORE_0 = 75; // -
    // int ASTORE_1 = 76; // -
    // int ASTORE_2 = 77; // -
    // int ASTORE_3 = 78; // -
    int IASTORE = 79; // visitInsn
    int LASTORE = 80; // -
    int FASTORE = 81; // -
    int DASTORE = 82; // -
    int AASTORE = 83; // -
    int BASTORE = 84; // -
    int CASTORE = 85; // -
    int SASTORE = 86; // -
    int POP = 87; // - X1 ->
    int POP2 = 88; // - X2 ->
    int DUP = 89; // -
    int DUP_X1 = 90; // -
    int DUP_X2 = 91; // - X2,Y1 -> Y1,X2,Y1
    int DUP2 = 92; // -
    int DUP2_X1 = 93; // - X1,Y2 -> Y2,X1,Y2
    int DUP2_X2 = 94; // - X2,Y2 -> Y2,X2,Y2
    int SWAP = 95; // - X1,Y1 -> Y1,X1
    int IADD = 96; // -
    int LADD = 97; // -
    int FADD = 98; // -
    int DADD = 99; // -
    int ISUB = 100; // -
    int LSUB = 101; // -
    int FSUB = 102; // -
    int DSUB = 103; // -
    int IMUL = 104; // -
    int LMUL = 105; // -
    int FMUL = 106; // -
    int DMUL = 107; // -
    int IDIV = 108; // -
    int LDIV = 109; // -
    int FDIV = 110; // -
    int DDIV = 111; // -
    int IREM = 112; // -
    int LREM = 113; // -
    int FREM = 114; // -
    int DREM = 115; // -
    int INEG = 116; // -
    int LNEG = 117; // -
    int FNEG = 118; // -
    int DNEG = 119; // -
    int ISHL = 120; // -
    int LSHL = 121; // -
    int ISHR = 122; // -
    int LSHR = 123; // -
    int IUSHR = 124; // -
    int LUSHR = 125; // -
    int IAND = 126; // -
    int LAND = 127; // -
    int IOR = 128; // -
    int LOR = 129; // -
    int IXOR = 130; // -
    int LXOR = 131; // -
    int IINC = 132; // visitIincInsn
    int I2L = 133; // visitInsn
    int I2F = 134; // -
    int I2D = 135; // -
    int L2I = 136; // -
    int L2F = 137; // -
    int L2D = 138; // -
    int F2I = 139; // -
    int F2L = 140; // -
    int F2D = 141; // -
    int D2I = 142; // -
    int D2L = 143; // -
    int D2F = 144; // -
    int I2B = 145; // -
    int I2C = 146; // -
    int I2S = 147; // -
    int LCMP = 148; // -
    int FCMPL = 149; // -
    int FCMPG = 150; // -
    int DCMPL = 151; // -
    int DCMPG = 152; // -
    int IFEQ = 153; // visitJumpInsn
    int IFNE = 154; // -
    int IFLT = 155; // -
    int IFGE = 156; // -
    int IFGT = 157; // -
    int IFLE = 158; // -
    int IF_ICMPEQ = 159; // -
    int IF_ICMPNE = 160; // -
    int IF_ICMPLT = 161; // -
    int IF_ICMPGE = 162; // -
    int IF_ICMPGT = 163; // -
    int IF_ICMPLE = 164; // -
    int IF_ACMPEQ = 165; // -
    int IF_ACMPNE = 166; // -
    int GOTO = 167; // -
    int JSR = 168; // -
    int RET = 169; // visitVarInsn
    int TABLESWITCH = 170; // visiTableSwitchInsn
    int LOOKUPSWITCH = 171; // visitLookupSwitch
    int IRETURN = 172; // visitInsn
    int LRETURN = 173; // -
    int FRETURN = 174; // -
    int DRETURN = 175; // -
    int ARETURN = 176; // -
    int RETURN = 177; // -
    int GETSTATIC = 178; // visitFieldInsn
    int PUTSTATIC = 179; // -
    int GETFIELD = 180; // -
    int PUTFIELD = 181; // -
    int INVOKEVIRTUAL = 182; // visitMethodInsn
    int INVOKESPECIAL = 183; // -
    int INVOKESTATIC = 184; // -
    int INVOKEINTERFACE = 185; // -
    int INVOKEDYNAMIC = 186; // visitInvokeDynamicInsn
    int NEW = 187; // visitTypeInsn
    int NEWARRAY = 188; // visitIntInsn
    int ANEWARRAY = 189; // visitTypeInsn
    int ARRAYLENGTH = 190; // visitInsn
    int ATHROW = 191; // -
    int CHECKCAST = 192; // visitTypeInsn
    int INSTANCEOF = 193; // -
    int MONITORENTER = 194; // visitInsn
    int MONITOREXIT = 195; // -
    // int WIDE = 196; // NOT VISITED
    int MULTIANEWARRAY = 197; // visitMultiANewArrayInsn
    int IFNULL = 198; // visitJumpInsn
    int IFNONNULL = 199; // -
    // int GOTO_W = 200; // -
    // int JSR_W = 201; // -

    // ## sunsong
    String CONSTRUCTOR_NAME = "<init>";
    String STATIC_NAME = "<clinit>";
    int COMPUTE_AUTO = 1 | 2;
    
    int ADD = Opcodes.IADD;
    int MUL = Opcodes.IMUL;
    int XOR = Opcodes.IXOR;
    int USHR = Opcodes.IUSHR;
    int SUB = Opcodes.ISUB;
    int DIV = Opcodes.IDIV;
    int NEG = Opcodes.INEG;
    int REM = Opcodes.IREM;
    int AND = Opcodes.IAND;
    int OR = Opcodes.IOR;

    int GT = Opcodes.IFGT;
    int LT = Opcodes.IFLT;
    int GE = Opcodes.IFGE;
    int LE = Opcodes.IFLE;
    int NE = Opcodes.IFNE;
    int EQ = Opcodes.IFEQ;
    
    Type T_void = Type.getType(void.class);
    Type T_boolean = Type.getType(boolean.class);
    Type T_byte = Type.getType(byte.class);
    Type T_char = Type.getType(char.class);
    Type T_short = Type.getType(short.class);
    Type T_int = Type.getType(int.class);
    Type T_float = Type.getType(float.class);
    Type T_long = Type.getType(long.class);
    Type T_double = Type.getType(double.class);
    
    Type T_Void = Type.getType(Void.class);
    Type T_Boolean = Type.getType(Boolean.class);
    Type T_Byte = Type.getType(Byte.class);
    Type T_Character = Type.getType(Character.class);
    Type T_Short = Type.getType(Short.class);
    Type T_Integer = Type.getType(Integer.class);
    Type T_Float = Type.getType(Float.class);
    Type T_Long = Type.getType(Long.class);
    Type T_Double = Type.getType(Double.class);
    Type T_Number = Type.getType(Number.class);
    Type T_Comparable = Type.getType(Comparable.class);
    Type T_BigDecimal = Type.getType(BigDecimal.class);
    
    Type T_Object = Type.getType(Object.class);
    Type T_String = Type.getType(String.class);
    Type T_Class = Type.getType(Class.class);
    Type T_Throwable = Type.getType(Throwable.class);
    Type T_StringBuilder = Type.getType(StringBuilder.class);
    Type T_Pattern = Type.getType(Pattern.class);
    Type T_List = Type.getType(List.class);
    Type T_Map = Type.getType(Map.class);
    Type T_Iterable = Type.getType(Iterable.class);
    Type T_Iterator = Type.getType(Iterator.class);
    Type T_Annotation = Type.getType(Annotation.class);
    Type T_Field = Type.getType(Field.class);
    Type T_Method = Type.getType(Method.class);

}
